baseapp/baseapp.go

```diff
@@ -534,53 +538,74 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte, tx sdk.Tx) (result sdk
 		ctx = ctx.WithMultiStore(msCache)
 	}
 
-	finalResult := sdk.Result{}
+	// accumulate results
 	var logs []string
-	for i, msg := range msgs {
+	var data []byte
+	var tags sdk.Tags
+	var code sdk.ABCICodeType
+	for msgIdx, msg := range msgs {
 		// Match route.
 		msgType := msg.Type()
 		handler := app.router.Route(msgType)
 		if handler == nil {
 			return sdk.ErrUnknownRequest("Unrecognized Msg type: " + msgType).Result()
 		}
 
-		result = handler(ctx, msg)
+		msgResult := handler(ctx, msg)
 
-		// Set gas utilized
-		finalResult.GasUsed += ctx.GasMeter().GasConsumed()
-		finalResult.GasWanted += result.GasWanted
+		// NOTE: GasWanted is determined by ante handler and
+		// GasUsed by the GasMeter
 
 		// Append Data and Tags
-		finalResult.Data = append(finalResult.Data, result.Data...)
-		finalResult.Tags = append(finalResult.Tags, result.Tags...)
-
-		// Construct usable logs in multi-message transactions. Messages are 1-indexed in logs.
-		logs = append(logs, fmt.Sprintf("Msg %d: %s", i+1, finalResult.Log))
+		data = append(data, msgResult.Data...)
+		tags = append(tags, msgResult.Tags...)
 
 		// Stop execution and return on first failed message.
-		if !result.IsOK() {
-			if len(msgs) == 1 {
-				return result
-			}
-			result.GasUsed = finalResult.GasUsed
-			if i == 0 {
-				result.Log = fmt.Sprintf("Msg 1 failed: %s", result.Log)
-			} else {
-				result.Log = fmt.Sprintf("Msg 1-%d Passed. Msg %d failed: %s", i, i+1, result.Log)
-			}
-			return result
+		if !msgResult.IsOK() {
+			logs = append(logs, fmt.Sprintf("Msg %d failed: %s", msgIdx, msgResult.Log))
+			code = msgResult.Code
+			break
 		}
+
+		// Construct usable logs in multi-message transactions.
+		logs = append(logs, fmt.Sprintf("Msg %d: %s", msgIdx, msgResult.Log))
+	}
+
+	// Set the final gas values.
+	result = sdk.Result{
+		Code:      code,
+		Data:      data,
+		Log:       strings.Join(logs, "\n"),
+		GasWanted: gasWanted,
+		GasUsed:   ctx.GasMeter().GasConsumed(),
+		// TODO: FeeAmount/FeeDenom
+		Tags: tags,
 	}
 
-	// If not a simulated run and result was successful, write to app.checkState.ms or app.deliverState.ms
-	// Only update state if all messages pass.
-	if mode != runTxModeSimulate && result.IsOK() {
+	// Only update state if all messages pass and we're not in a simulation.
+	if result.IsOK() && mode != runTxModeSimulate {
 		msCache.Write()
 	}
 
-	finalResult.Log = strings.Join(logs, "\n")
+	return result
+}
+
+func (app *BaseApp) getContextForAnte(mode runTxMode, txBytes []byte) sdk.Context {
+	var ctx sdk.Context
+
+	// Get the context.
+	if mode == runTxModeCheck || mode == runTxModeSimulate {
+		ctx = app.checkState.ctx.WithTxBytes(txBytes)
+	} else {
+		ctx = app.deliverState.ctx.WithTxBytes(txBytes)
+		ctx = ctx.WithSigningValidators(app.signedValidators)
```

Comment:
Maybe I'm unclear on what the purpose of simulate is, but why doesn't it get access to any of the app's signed validators? (How can it simulate how a slashing or staking tx will happen w/o this?)

Reply:
Yeh good question. The problem is that information is only supplied by BeginBlock, and when we're simulating, we don't know what the next BeginBlock will look like ...

Reply:
Hrmm. Would it make sense to have an option where its simulated as the last tx in the last known block (for things that depend on headers / validator set), and an option for on the next block (for conceivable things that depend on the exact block height)? This discussion should probably be carried out in a separate issue though

Reply:
Could be useful, although that seems like a rare case - all slashing logic related to signing (downtime / evidence) happens in a `BeginBlock` handler, not in transactions.